File: engines/resilienceguard.py
"""
ResilienceGuard Adaptive Disruption Engine
Pure Python equivalent to src/engines/resilienceguard/resilienceguard.ts
Handles dockside canister rejections (leaks, illegible labels) and institutional lab withdrawals,
automatically computing replacement strategies and promoting standby containers to preserve the >=150L hauler threshold.
"""

from datetime import datetime
from typing import List, Optional, Tuple, Dict, Any

from models import PickupLot, WasteItem, DisruptionLog, RecoveryPlan, DisruptionType
from engines.chemiguard import ChemiGuardEngine


class ResilienceGuardEngine:
    """
    Autonomous self-healing engine for supply-chain and dockside disruptions.
    """

    @classmethod
    def plan_lab_cancellation_recovery(
        cls,
        lot: PickupLot,
        cancelled_lab_id: str,
        current_lot_items: List[WasteItem],
        available_pool: List[WasteItem],
    ) -> RecoveryPlan:
        """
        Evaluates the impact of a participating laboratory abruptly cancelling its pickup commitment.
        """
        cancelled_items = [i for i in current_lot_items if i.lab_id == cancelled_lab_id]
        lost_volume = sum(i.volume_liters for i in cancelled_items)
        remaining_items = [i for i in current_lot_items if i.lab_id != cancelled_lab_id]
        remaining_volume = sum(i.volume_liters for i in remaining_items)

        is_below_threshold = remaining_volume < lot.target_threshold_liters
        replacement_items: List[WasteItem] = []
        recovered_volume = 0.0

        if is_below_threshold:
            # Search candidate standby containers from non-cancelled labs
            candidates = [
                item for item in available_pool
                if item.lab_id != cancelled_lab_id
                and not any(curr.id == item.id for curr in current_lot_items)
                and item.status != 'IN_LOT'
                and item.condition != 'LEAKING'
                and all(ChemiGuardEngine.check_pair(item, rem) is None for rem in remaining_items)
            ]

            # Sort candidates by volume descending
            candidates.sort(key=lambda x: x.volume_liters, reverse=True)

            for candidate in candidates:
                # Check compatibility with already added replacements
                is_compatible_with_replacements = all(
                    ChemiGuardEngine.check_pair(candidate, rep) is None
                    for rep in replacement_items
                )

                if is_compatible_with_replacements:
                    replacement_items.append(candidate)
                    recovered_volume += candidate.volume_liters

                    if remaining_volume + recovered_volume >= lot.target_threshold_liters:
                        break

        new_projected_volume = remaining_volume + recovered_volume
        can_be_fully_restored = (not is_below_threshold) or (new_projected_volume >= lot.target_threshold_liters)

        if can_be_fully_restored:
            if is_below_threshold:
                explanation = (
                    f"ResilienceGuard identified {len(replacement_items)} compatible standby container(s) "
                    f"(+{recovered_volume:.1f}L). Quota will be restored to {new_projected_volume:.1f}L "
                    f"(>= {lot.target_threshold_liters:.1f}L)."
                )
            else:
                explanation = (
                    f"Lot volume remains above quota ({remaining_volume:.1f}L >= {lot.target_threshold_liters:.1f}L) "
                    f"thanks to the initial reserve safety buffer."
                )
        else:
            explanation = (
                f"Insufficient compatible standby waste available in region (+{recovered_volume:.1f}L found). "
                f"Lot projected at {new_projected_volume:.1f}L, which is under the {lot.target_threshold_liters:.1f}L "
                f"hauler quota. Additional lab pooling required."
            )

        return RecoveryPlan(
            lot_id=lot.id,
            disruption_type='LAB_CANCELLATION',
            description=f"Lab #{cancelled_lab_id} withdrew {len(cancelled_items)} container(s) totaling {lost_volume:.1f} L.",
            affected_volume_liters=lost_volume,
            remaining_volume_before_recovery=remaining_volume,
            threshold_liters=lot.target_threshold_liters,
            is_below_threshold=is_below_threshold,
            can_be_fully_restored=can_be_fully_restored,
            replacement_item_ids=[i.id for i in replacement_items],
            replacement_items=replacement_items,
            new_projected_volume=new_projected_volume,
            explanation=explanation,
        )

    @classmethod
    def plan_canister_rejection_recovery(
        cls,
        lot: PickupLot,
        rejected_item: WasteItem,
        reason: str,
        current_lot_items: List[WasteItem],
        available_pool: List[WasteItem],
    ) -> RecoveryPlan:
        """
        Evaluates partial hauler rejection at the loading dock (e.g. leaking drum or unreadable label).
        Promotes pre-approved standby containers to prevent lot cancellation.
        """
        remaining_items = [i for i in current_lot_items if i.id != rejected_item.id]
        remaining_volume = sum(i.volume_liters for i in remaining_items)
        is_below_threshold = remaining_volume < lot.target_threshold_liters

        replacement_items: List[WasteItem] = []
        recovered_volume = 0.0

        if is_below_threshold:
            # Look first in lot's pre-approved standby containers / available pool
            standby_candidates = [
                item for item in available_pool
                if item.id != rejected_item.id
                and item.condition != 'LEAKING'
                and not any(curr.id == item.id for curr in current_lot_items)
                and all(ChemiGuardEngine.check_pair(item, rem) is None for rem in remaining_items)
            ]

            for candidate in standby_candidates:
                replacement_items.append(candidate)
                recovered_volume += candidate.volume_liters
                if remaining_volume + recovered_volume >= lot.target_threshold_liters:
                    break

        new_projected_volume = remaining_volume + recovered_volume
        can_be_fully_restored = (not is_below_threshold) or (new_projected_volume >= lot.target_threshold_liters)

        if is_below_threshold:
            if can_be_fully_restored:
                explanation = (
                    f"Volume fell to {remaining_volume:.1f}L (< {lot.target_threshold_liters:.1f}L). "
                    f"ResilienceGuard promoted standby container (+{recovered_volume:.1f}L) to restore hauler dispatch."
                )
            else:
                explanation = (
                    f"Lot dropped to {remaining_volume:.1f}L (< {lot.target_threshold_liters:.1f}L) "
                    f"and insufficient compatible standby containers were available."
                )
        else:
            explanation = (
                f"Remaining lot volume is {remaining_volume:.1f}L, safely protected above the "
                f"{lot.target_threshold_liters:.1f}L quota by the initial reserve buffer."
            )

        return RecoveryPlan(
            lot_id=lot.id,
            disruption_type='CONTAINER_REJECTED',
            description=f"Hauler driver rejected {rejected_item.name} ({rejected_item.volume_liters:.1f}L) due to {reason}.",
            affected_volume_liters=rejected_item.volume_liters,
            remaining_volume_before_recovery=remaining_volume,
            threshold_liters=lot.target_threshold_liters,
            is_below_threshold=is_below_threshold,
            can_be_fully_restored=can_be_fully_restored,
            replacement_item_ids=[i.id for i in replacement_items],
            replacement_items=replacement_items,
            new_projected_volume=new_projected_volume,
            explanation=explanation,
        )

File: src/engines/resilienceguard/resilienceguard.ts
import type { PickupLot, WasteItem, DisruptionLog } from '../../types';
import { ChemiGuardEngine } from '../chemiguard/chemiguard';
import { db } from '../../storage/db';

export interface RecoveryPlan {
  lotId: string;
  disruptionType: 'LAB_CANCELLATION' | 'CONTAINER_REJECTED' | 'CONTAINER_DAMAGED';
  description: string;
  affectedVolumeLiters: number;
  remainingVolumeBeforeRecovery: number;
  thresholdLiters: number;
  isBelowThreshold: boolean;
  canBeFullyRestored: boolean;
  replacementItemIds: string[];
  replacementItems: WasteItem[];
  newProjectedVolume: number;
  explanation: string;
}

export class ResilienceGuardEngine {
  /**
   * Evaluates impact of a lab cancelling its pickup commitment.
   */
  public static planLabCancellationRecovery(
    lot: PickupLot,
    cancelledLabId: string,
    currentLotItems: WasteItem[],
    availablePool: WasteItem[]
  ): RecoveryPlan {
    const cancelledItems = currentLotItems.filter((i) => i.labId === cancelledLabId);
    const lostVolume = cancelledItems.reduce((s, i) => s + i.volumeLiters, 0);
    const remainingItems = currentLotItems.filter((i) => i.labId !== cancelledLabId);
    const remainingVolume = remainingItems.reduce((s, i) => s + i.volumeLiters, 0);

    const isBelowThreshold = remainingVolume < lot.targetThresholdLiters;
    const replacementItems: WasteItem[] = [];
    let recoveredVolume = 0;

    if (isBelowThreshold) {
      // 1. Search candidate standby containers from non-cancelled labs
      const candidates = availablePool.filter((item) => {
        if (item.labId === cancelledLabId) return false;
        if (currentLotItems.some((curr) => curr.id === item.id)) return false;
        if (item.status === 'IN_LOT' || item.condition === 'LEAKING') return false;

        // Must be compatible with all remaining items in the lot
        return remainingItems.every((rem) => ChemiGuardEngine.checkPair(item, rem) === null);
      });

      // Sort candidates by urgency, then volume
      candidates.sort((a, b) => b.volumeLiters - a.volumeLiters);

      for (const candidate of candidates) {
        // Check compatibility with already added replacements
        const isCompatibleWithReplacements = replacementItems.every(
          (rep) => ChemiGuardEngine.checkPair(candidate, rep) === null
        );

        if (isCompatibleWithReplacements) {
          replacementItems.push(candidate);
          recoveredVolume += candidate.volumeLiters;

          if (remainingVolume + recoveredVolume >= lot.targetThresholdLiters) {
            break;
          }
        }
      }
    }

    const newProjectedVolume = remainingVolume + recoveredVolume;
    const canBeFullyRestored = !isBelowThreshold || newProjectedVolume >= lot.targetThresholdLiters;

    return {
      lotId: lot.id,
      disruptionType: 'LAB_CANCELLATION',
      description: `Lab #${cancelledLabId} withdrew ${cancelledItems.length} container(s) totaling ${lostVolume.toFixed(1)} L.`,
      affectedVolumeLiters: lostVolume,
      remainingVolumeBeforeRecovery: remainingVolume,
      thresholdLiters: lot.targetThresholdLiters,
      isBelowThreshold,
      canBeFullyRestored,
      replacementItemIds: replacementItems.map((i) => i.id),
      replacementItems,
      newProjectedVolume,
      explanation: canBeFullyRestored
        ? isBelowThreshold
          ? `ResilienceGuard identified ${replacementItems.length} compatible standby container(s) (+${recoveredVolume.toFixed(1)}L). Quota will be restored to ${newProjectedVolume.toFixed(1)}L (>= ${lot.targetThresholdLiters}L).`
          : `Lot volume remains above quota (${remainingVolume.toFixed(1)}L >= ${lot.targetThresholdLiters}L) thanks to the initial reserve safety buffer.`
        : `Insufficient compatible standby waste available in region (+${recoveredVolume.toFixed(1)}L found). Lot projected at ${newProjectedVolume.toFixed(1)}L, which is under the ${lot.targetThresholdLiters}L hauler quota. Additional lab pooling required.`,
    };
  }

  /**
   * Evaluates partial hauler rejection at the loading dock (e.g. leaking drum or unreadable label).
   */
  public static planCanisterRejectionRecovery(
    lot: PickupLot,
    rejectedItem: WasteItem,
    reason: 'DAMAGED_LEAKING' | 'LABEL_ILLEGIBLE' | 'CONTAINER_BULGING',
    currentLotItems: WasteItem[],
    availablePool: WasteItem[]
  ): RecoveryPlan {
    const remainingItems = currentLotItems.filter((i) => i.id !== rejectedItem.id);
    const remainingVolume = remainingItems.reduce((s, i) => s + i.volumeLiters, 0);
    const isBelowThreshold = remainingVolume < lot.targetThresholdLiters;

    const replacementItems: WasteItem[] = [];
    let recoveredVolume = 0;

    if (isBelowThreshold) {
      // Look first in lot's pre-approved standby containers
      const standbyCandidates = availablePool.filter((item) => {
        if (item.id === rejectedItem.id) return false;
        if (item.condition === 'LEAKING') return false;
        if (currentLotItems.some((curr) => curr.id === item.id)) return false;

        return remainingItems.every((rem) => ChemiGuardEngine.checkPair(item, rem) === null);
      });

      for (const candidate of standbyCandidates) {
        replacementItems.push(candidate);
        recoveredVolume += candidate.volumeLiters;
        if (remainingVolume + recoveredVolume >= lot.targetThresholdLiters) break;
      }
    }

    const newProjectedVolume = remainingVolume + recoveredVolume;
    const canBeFullyRestored = !isBelowThreshold || newProjectedVolume >= lot.targetThresholdLiters;

    return {
      lotId: lot.id,
      disruptionType: 'CONTAINER_REJECTED',
      description: `Hauler driver rejected ${rejectedItem.name} (${rejectedItem.volumeLiters.toFixed(1)}L) due to ${reason}.`,
      affectedVolumeLiters: rejectedItem.volumeLiters,
      remainingVolumeBeforeRecovery: remainingVolume,
      thresholdLiters: lot.targetThresholdLiters,
      isBelowThreshold,
      canBeFullyRestored,
      replacementItemIds: replacementItems.map((i) => i.id),
      replacementItems,
      newProjectedVolume,
      explanation: isBelowThreshold
        ? canBeFullyRestored
          ? `Volume fell to ${remainingVolume.toFixed(1)}L (< ${lot.targetThresholdLiters}L). ResilienceGuard promoted standby container (+${recoveredVolume.toFixed(1)}L) to restore hauler dispatch.`
          : `Lot dropped to ${remainingVolume.toFixed(1)}L (< ${lot.targetThresholdLiters}L) and insufficient compatible standby containers were available.`
        : `Remaining lot volume is ${remainingVolume.toFixed(1)}L, safely protected above the ${lot.targetThresholdLiters}L quota by the initial reserve buffer.`,
    };
  }

  /**
   * Executes the recovery plan atomically in Dexie IndexedDB.
   */
  public static async executeRecoveryPlan(
    lot: PickupLot,
    plan: RecoveryPlan,
    offendingItemIdsToRemove: string[]
  ): Promise<void> {
    await db.transaction('rw', [db.pickupLots, db.wasteItems, db.disruptionLogs], async () => {
      // 1. Remove cancelled or rejected items from the lot
      for (const id of offendingItemIdsToRemove) {
        await db.wasteItems.update(id, {
          status: plan.disruptionType === 'CONTAINER_REJECTED' ? 'QUARANTINED' : 'AVAILABLE',
          assignedLotId: undefined,
        });
      }

      // 2. Add replacement items to the lot
      for (const id of plan.replacementItemIds) {
        await db.wasteItems.update(id, {
          status: 'IN_LOT',
          assignedLotId: lot.id,
        });
      }

      // 3. Update the lot record
      const updatedItemIds = lot.wasteItemIds
        .filter((id) => !offendingItemIdsToRemove.includes(id))
        .concat(plan.replacementItemIds);

      await db.pickupLots.update(lot.id, {
        wasteItemIds: updatedItemIds,
        totalVolumeLiters: plan.newProjectedVolume,
        isThresholdMet: plan.newProjectedVolume >= lot.targetThresholdLiters,
        rejectionIncidentCount: (lot.rejectionIncidentCount || 0) + 1,
      });

      // 4. Record disruption log
      const log: DisruptionLog = {
        id: `disrupt-${Date.now()}`,
        timestamp: new Date().toISOString(),
        type: plan.disruptionType,
        description: plan.description,
        recoveryActionTaken: plan.explanation,
        recoveredVolumeDelta: plan.replacementItems.reduce((s, i) => s + i.volumeLiters, 0),
      };
      await db.disruptionLogs.add(log);
    });
  }
}
